File: csharp/tree-building/TreeBuilding.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class TreeBuildingRecord
{
    public int ParentId { get; set; }
    public int RecordId { get; set; }
}

public class Tree
{
    public Tree(int id)
    {
        Id = id;
        Children = new List<Tree>();
    }

    public int Id { get; }

    public List<Tree> Children { get; }

    public bool IsLeaf => Children.Count == 0;
}

public static class TreeBuilder
{
    private const int RootRecordId = 0;

    public static Tree BuildTree(IEnumerable<TreeBuildingRecord> records)
    {
        var subtrees = new List<Tree>();

        var expectedRecordId = RootRecordId;
        foreach (var record in records.OrderBy(r => r.RecordId))
        {
            ValidateRecord(record, expectedRecordId);

            subtrees.Add(new Tree(record.RecordId));

            if (!record.IsRoot())
            {
                subtrees[record.ParentId].Children.Add(subtrees[record.RecordId]);
            }

            expectedRecordId++;
        }

        if (subtrees.Count == 0)
        {
            throw new ArgumentException("Cannot build tree from empty records");
        }

        return subtrees[RootRecordId];
    }

    private static void ValidateRecord(TreeBuildingRecord record, int expectedRecordId)
    {
        if (record.RecordId != expectedRecordId)
        {
            throw new ArgumentException($"Record {record.RecordId} has invalid identifier");
        }

        if (record.IsRoot() && record.ParentId != record.RecordId)
        {
            throw new ArgumentException($"Root record {record.RecordId} has invalid parent identifier");
        }

        if (!record.IsRoot() && record.ParentId >= record.RecordId)
        {
            throw new ArgumentException($"Record {record.RecordId} has invalid parent identifier");
        }
    }

    private static bool IsRoot(this TreeBuildingRecord record) => record.RecordId == RootRecordId;
}

File: fsharp/tree-building/TreeBuilding.fs
// This is the file you need to modify for your own solution.
// The unit tests will use this code, and it will be used by the benchmark tests
// for the "Mine" row of the summary table.

// Remember to not only run the unit tests for this exercise, but also the
// benchmark tests using `dotnet run -c Release`.
// Please refer to the instructions for more information about the benchmark tests.

module TreeBuilding

open TreeBuildingTypes

type Tree =
    | Branch of int * Tree list
    | Leaf of int

let recordId t =
    match t with
    | Branch (id, c) -> id
    | Leaf id -> id

let isBranch t =
    match t with
    | Branch (id, c) -> true
    | Leaf id -> false

let children t =
    match t with
    | Branch (id, c) -> c
    | Leaf id -> []

let private isChildOf child parent = child.ParentId = parent.RecordId

let rec buildSubtree records root =
    match List.filter (isChildOf root) records with
    | [] -> Leaf root.RecordId
    | xs ->
        let children =
            xs
            |> List.map (buildSubtree records)
            |> List.sortBy recordId

        Branch(root.RecordId, children)

let buildTree records =
    match List.sortBy (fun r -> r.RecordId) records with
    | [] -> failwith "Cannot build tree from empty records"
    | xs when
        xs
        |> List.indexed
        |> List.exists (fun (i, r) -> r.RecordId <> i) -> failwith "Invalid record identifiers found"
    | _ :: children when
        children
        |> List.exists (fun r -> r.ParentId >= r.RecordId) -> failwith "Invalid parent identifiers found"
    | root :: _ when isChildOf root root |> not -> failwith "Invalid root node"
    | root :: children -> buildSubtree children root
